pkg/icingadb/ha.go

```diff
@@ -223,103 +224,117 @@ func (h *HA) controller() {
+						return errors.Wrap(err, "can't scan row")
+					}
+
+					if shouldLog {
+						h.logger.Infow("Another instance is active",
+							zap.String("instance_id", instance.Id.String()),
+							zap.String("environment", envId.String()),
+							"heartbeat", instance.Heartbeat,
+							zap.Duration("heartbeat_age", time.Since(instance.Heartbeat.Time())))
+					}
+				} else if err := rows.Err(); err != nil {
+					return errors.Wrap(err, "can't fetch row")
 				} else {
-					h.logger.Debugw("Can't update or insert instance. Retrying", zap.Error(err), zap.Int("retry count", attempt))
+					takeover = true
 				}
-				continue
+
+				return nil
+			}()
+			if errTakeover != nil {
+				return errTakeover
 			}
-		}
 
-		if err := tx.Commit(); err != nil {
-			cancelCtx()
-			return errors.Wrap(err, "can't commit transaction")
-		}
+			i := v1.IcingadbInstance{
+				EntityWithoutChecksum: v1.EntityWithoutChecksum{
+					IdMeta: v1.IdMeta{
+						Id: h.instanceId,
+					},
+				},
+				EnvironmentMeta: v1.EnvironmentMeta{
+					EnvironmentId: envId,
+				},
+				Heartbeat:                         *t,
+				Responsible:                       types.Bool{Bool: takeover || h.responsible, Valid: true},
+				EndpointId:                        s.EndpointId,
+				Icinga2Version:                    s.Version,
+				Icinga2StartTime:                  s.ProgramStart,
+				Icinga2NotificationsEnabled:       s.NotificationsEnabled,
+				Icinga2ActiveServiceChecksEnabled: s.ActiveServiceChecksEnabled,
+				Icinga2ActiveHostChecksEnabled:    s.ActiveHostChecksEnabled,
+				Icinga2EventHandlersEnabled:       s.EventHandlersEnabled,
+				Icinga2FlapDetectionEnabled:       s.FlapDetectionEnabled,
+				Icinga2PerformanceDataEnabled:     s.PerformanceDataEnabled,
+			}
 
-		if takeover {
-			// Insert the environment after each heartbeat takeover if it does not already exist in the database
-			// as the environment may have changed, although this is likely to happen very rarely.
-			if err := h.insertEnvironment(); err != nil {
-				cancelCtx()
-				return errors.Wrap(err, "can't insert environment")
+			stmt, _ := h.db.BuildUpsertStmt(i)
+			if _, err := tx.NamedExecContext(ctx, stmt, i); err != nil {
+				return internal.CantPerformQuery(err, stmt)
 			}
 
-			h.signalTakeover()
+			if err := tx.Commit(); err != nil {
+				return errors.Wrap(err, "can't commit transaction")
+			}
+
+			return nil
+		},
+		IsRetryable,
+		backoff.NewExponentialWithJitter(time.Millisecond*256, time.Second*3),
+		retry.Settings{
+			OnError: func(_ time.Duration, attempt uint64, err, lastErr error) {
+				if lastErr == nil || err.Error() != lastErr.Error() {
+					log := h.logger.Debugw
+					if attempt > 2 {
+						log = h.logger.Infow
+					}
+
+					log("Can't update or insert instance. Retrying", zap.Error(err), zap.Uint64("retry count", attempt))
+				}
+			},
+		},
+	)
+	if err != nil {
+		h.logger.Errorw("Can't update or insert instance", zap.Error(err))
+		return err
```

Comment:
When `realize` returns with an error it will be logged anyway. So this could be simplified to just returning `errors.Wrap()`.